samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java

```diff
@@ -19,36 +19,201 @@
+        throw new SamzaException("No jobs to run.");
+      }
+      plan.getJobConfigs().forEach(jobConfig -> {
+        log.info("Starting job {} StreamProcessor with config {}", jobConfig.getName(), jobConfig);
+
+        StreamProcessor processor = createStreamProcessor(jobConfig, app);
+        processor.addLifeCycleAware(new LocalProcessorListener(processor));
+        processors.add(processor);
+        processor.start();
+      });
+      appStatus = ApplicationStatus.Running;
+
+      // 4. block until the processors are done or there is a failure
+      awaitComplete();
+
+    } catch (Throwable t) {
+      throw new SamzaException("Failed to run application", t);
+    } finally {
+      if (coordination != null) {
+        coordination.reset();
+      }
+    }
   }
 
   @Override
   public void kill(StreamApplication streamApp) {
+    processors.forEach(StreamProcessor::stop);
   }
 
   @Override
   public ApplicationStatus status(StreamApplication streamApp) {
-    return null;
+    return appStatus;
+  }
+
+  /**
+   * Create the Coordination needed by the application runner.
+   * @return {@link CoordinationUtils}
+   * @throws Exception exception
+   */
+  private CoordinationUtils getCoordinationUtils() throws Exception {
+    ApplicationConfig appConfig = new ApplicationConfig(config);
+    String clazz = appConfig.getCoordinationServiceFactoryClass();
+    if (clazz != null) {
+      CoordinationServiceFactory factory = ClassLoaderHelper.fromClassName(clazz);
+      String groupId = String.format("app-%s-%s", config.get(JobConfig.JOB_NAME()), config.get(JobConfig.JOB_ID(), "1"));
+      return factory.getCoordinationService(groupId, String.valueOf(processorId), config);
+    }
+    else {
+      return null;
+    }
+  }
+
+  /**
+   * Create intermediate streams.
+   * @param intStreams intermediate {@link StreamSpec}s
+   * @throws Exception exception
+   */
+  /* package private */ void createStreams(List<StreamSpec> intStreams) throws Exception {
+    if (!intStreams.isEmpty()) {
+      if (coordination != null) {
+        Latch initLatch = coordination.getLatch(1, LATCH_INIT);
+        coordination.getLeaderElector().tryBecomeLeader(() -> {
+          getStreamManager().createStreams(intStreams);
+          initLatch.countDown();
+        });
+        initLatch.await(LATCH_TIMEOUT, TimeUnit.MINUTES);
+      } else {
+        // each application process will try creating the streams, which
+        // requires stream creation to be idempotent
+        getStreamManager().createStreams(intStreams);
+      }
+    }
+  }
+
+  /**
+   * Create {@link StreamProcessor} based on {@link StreamApplication} and the config
+   * @param config config
+   * @param app {@link StreamApplication}
+   * @return {@link StreamProcessor]}
+   */
+  /* package private */ StreamProcessor createStreamProcessor(Config config, StreamApplication app) {
+    Object taskFactory = TaskFactoryUtil.createTaskFactory(config, app, this);
```

Comment:
side note again. This makes me feel that it is still worth to use a common generic type for TaskFactory, at least for the TaskFactory object itself to be bound to a defined interface instead of an Object. Granted that we are going to load the TaskFactory class and lose the type info of the task object returned from createInstance(). But we at least know the type of the TaskFactory object and have a defined method createInstance() to be called without the need to type cast the factory object to StreamTaskFactory/AsyncStreamTaskFactory before we can even invoke the createInstance() method.
@navina and @prateekm, thoughts?